### DISCUSSION TITLE
"Could not load scales config" error when running zwavejs from an esbuild bundle

### DISCUSSION THREAD

**Discussion by hspaay** [zwave-js/zwave-js#5592] [Q&A] [answered]
Hi, I'm hitting a roadblock building an app that uses zwave-js. The app runs fine when running with node after compiling from typescript but when bundling with esbuild, a bunch of errors pop-up:

Running the app:
$ZWAVEJS_EXTERNAL_CONFIG=/home/user/bin/hiveot/stores/zwavejs/configdb dist/app.cjs

Results in:
> Could not load scales config: The device classes config file does not exist! (ZW0340)
> 15:02:49.126 CONFIG   Could not load manufacturers config: The manufacturer config file does not exist! (ZW0340)
> Writing the device index to disk failed: ENOENT: no such file or directory, open '/home/user/hiveot/stores/zwavejs/configdb/devices/index.json'
> 15:02:49.128 CONFIG   Could not load notifications config: The config file does not exist! (ZW0340)
> 15:02:49.129 CONFIG   Could not load scales config: The named scales config file does not exist! (ZW0340)
> 15:02:49.129 CONFIG   Could not load sensor types config: The sensor types config file does not exist! (ZW0340)
> 15:02:49.130 CONFIG   Could not meters config: The config file does not exist! (ZW0340)
> 15:02:49.130 CONFIG   Could not load indicators config: The config file does not exist! (ZW0340)

The app continues to run which is very encouraging, but obviously there device models are not loaded and there is no zwave metadata for the discovered node. 

The esbuild command used:
> esbuild src/main.ts --platform=node  --outfile=dist/app.cjs --external:@serialport --bundle --loader:.ts=ts
The resulting binary is only 5.9MB which is less than I expected.

My guess is that the esbuild does not pull in the device models as these are dynamically loaded. So the question is how to included the device models with esbuild so that zwave-js can load them, or alternatively can they be loaded externally?

It also seems like the ZWAVEJS_EXTERNAL_CONFIG isn't working but it does work when running node with the app without esbuild. 

Any pointers in the right direction would be much appreciated.
Cheers,

ps: a search for similar problems lead me to solve another problem and add --external:@serialport. I can't find any documentation on bundling a zwave-js application with esbuild however.
pps: in case the question comes up, why do you need esbuild, that is to be able to run 'pkg' so the app can be installed as a standalone executable.

**Comment by AlCalzone** [comment]
You will have to distribute the contents of [this folder](https://github.com/zwave-js/node-zwave-js/tree/master/packages/config/config) along with the bundled app. Not sure about the location relative to the executable though.

What value are you setting `ZWAVEJS_EXTERNAL_CONFIG` to? It needs to be an absolute path.

**Comment by hspaay** [comment]
@AlCalzone thanks for the quick reply!

I've tried different values for ZWAVE_JS_EXTERNAL_CONFIG. It looks like this just needs to point to an empty folder  (yes, absolute path) and zwave-js will populate it. Is this just a cache folder?

Thanks for the hint on including the content of 'node_modules/@zwave-js/config/config' external to the bundle. I'm still trying to figure out where it is supposed to be relative to the bundle. 
When running from the ./dist folder itself that contains the bundle I've tried:
 config/
 config/config
 node_modules/@zwave-js/config/config

No luck sofar. 
Maybe best to find a way to log the folder location that it tries to open so I'm not looking in the dark.
Thanks for the help!

UPDATE
-----------
Its working (sort of) :)  Adding a log statement in the ConfigManager showed the full path.
When running from the './dist/' folder, the config folder must reside in the root of the project. Eg the folder used when building the bundle. 

Now to figure how how to make this portable to another folder...

**Comment by hspaay** [comment, marked as answer]
Just in case someone else has the same question, this is the summary:
(thanks @AlCalzone for pointing me in the right direction)

UPDATE: forget all the stuff below. Simply make sure you use 'CommonJS' and 'pkg' will work just fine.

---

## Objective
Build a zwave binding using node-zwave-js, go/wasm for talking to the backend, and bundle the whole thing as a single executable. (linux version)

## Application structure
```
   * src/         // the binding source code in typescript
   * cmd/         // if any, in my case this contains the wasm/golang code
   * build/       // hold the intermediate wasm, typescript and esbuild output
   * dist/        // final output
   * node_modules/   
```
## create the build configs
* [package.json](https://github.com/hiveot/bindings/blob/main/zwavejs/package.json)
* [tsconfig.json](https://github.com/hiveot/bindings/blob/main/zwavejs/tsconfig.json)
* [esbuild.config.mjs](https://github.com/hiveot/bindings/blob/main/zwavejs/esbuild.config.mjs)
* [go.mod](https://github.com/hiveot/bindings/blob/main/zwavejs/go.mod)

## Build the whole thing (in package.json scripts section)
```
 GOOS=js GOARCH=wasm go build -o build/hapi.wasm cmd/hapi/main.go
 tsc  
 node ./esbuild.config.mjs
 ln -sf node_modules/@zwave-js/config/config            
 pkg package.json
```
## Explanation

1. The first step is to build the wasm file. Too much to go into detail here but the interesting bit is that there is a wasm blob that ends up in the build directory and must be bundled in the application. One problem is that the environment exceeded 8MB which wasm_exec.js doesn't like, so I had to remove the worst offenders (in this case XDG variables). 
2. Compiling the app to typescript. Courtesy of [tsconfig.json](https://github.com/hiveot/bindings/blob/main/zwavejs/tsconfig.json) this is straightforward. This javascript code tree ends up in the build/ folder.
3. Bundle the javascript using esbuild into a commonjs app. This is needed because 'pkg' doesn't handle ES modules](https://github.com/vercel/pkg/issues/1291) yet. The end result is a single app.cjs file in the build/ directory.

5. Create a symlink from node_modules/@zwave-js/config/config to the project root. This is unfortunate but required as pkg changes "__dirname" in **exports.configDir = path_1.default.resolve(__dirname, "../config");** in config/build/utils.js, to 'snapshot/', which is the root of the image produced by pkg.

6. Run 'pkg package.json'.  package.json includes the configuration for 'pkg'. This is where the wasm file and the device config assets are included. 
Also included in the assets section is node_modules/@serialport as this is a native piece of code that the bundler doesn't handle. It is therefore excluded in esbuild.config.mjs and included as source in pkg.
 
7. ZWAVEJS_EXTERNAL_CONFIG must be set as zwave-js wants that directory writable. There  is possibly a workaround when messing with the version file but I didn't want to go there. 
Instead of setting this environment variable on the commandline, it is set in the application main.ts to the cache directory. E.g. in src/main.ts:
```
   process.env.ZWAVEJS_EXTERNAL_CONFIG = path.join(appConfig.cacheDir, "config")
```

## TODO
1. Build the wasm using tinygo to get it down from 5MB to <1MB.
2. Reduce the pkg size. The resulting binary is currently over 100MB! Yes, Ouch! 
3. Figure out what the build warnings mean and fix them.
4. Trim down the environment variables passed to wasm.
5. Make it portable to other platforms. pkg requires a platform target. So currently building for arm64 requires editing the package.json.

**Reply by AlCalzone** [reply]
I'm a bit confused about the mentioning of ES Modules. Zwave-js is still CommonJS, otherwise `__dirname` and similar wouldn't exist.

**Reply by hspaay** [reply]
Ah I see. The binding I've written uses ESM. This is why 'pkg' is unhappy. That said, I might change the binding to CommonJS if that avoids the esbuild step and fits better with zwave-js.  At times I'm still struggling to wrap my head around the whole ecosystem of typescript, javascript, commonjs, es5, es6, es2015, es2018, node version 9, 14, 18, esbuild and pkg. My (simplistic) thinking was the newer stuff improved interop and reduce the chance of running into problems.
Changing the binding to CommonJS probably won't change the __dirname issue and the need to put config in the project root for packaging, but I'm a big fan of solving problems by elimination instead of addition.
Thanks for pointing that out.

**Reply by AlCalzone** [reply]
> My (simplistic) thinking was the newer stuff improved interop and reduce the chance of running into problems.

Hehe. ESM is a good idea but a f*ing pain to work with right now.

**Reply by hspaay** [reply]
OMG! This is the answer.
With simply using CommonJS and removing esbuild, 'pkg' just works. It also embeds config without having to mention it in the assets section of pkg.  Size of the executable is much reduced as well. 58M with GZip compression.
So, basically, my post can be reduced to 'use Common~~Sense~~JS'
